server/neptune/workflows/internal/deploy/signals/revision.go

```diff
@@ -0,0 +1,77 @@
+package signals
+
+import (
+	"time"
+
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/deploy/revision/queue"
+	"go.temporal.io/sdk/workflow"
+)
+
+const NewRevisionID = "new-revision"
+
+type NewRevisionRequest struct {
+	Revision string
+}
+
+type Queue interface {
+	Push(queue.Message)
+}
+
+func NewRevisionSignal(ctx workflow.Context, queue Queue, timeout time.Duration) *NewRevision {
+	return &NewRevision{
+		input:           workflow.GetSignalChannel(ctx, NewRevisionID),
+		queue:           queue,
+		timeoutDuration: timeout,
+	}
+}
+
+type NewRevision struct {
+	input           workflow.ReceiveChannel
+	queue           Queue
+	timeoutDuration time.Duration
+
+	// mutable
+	timerCancel workflow.CancelFunc
+	timeout     bool
+}
+
+func (n *NewRevision) AddCallback(ctx workflow.Context, selector workflow.Selector) {
+	selector.AddReceive(n.input, func(c workflow.ReceiveChannel, more bool) {
+
+		// more is false when the channel is closed, so let's just return right away
+		if !more {
+			return
+		}
+
+		var request NewRevisionRequest
+		c.Receive(ctx, &request)
+
+		// cancel's the existing timeout timer
+		n.timerCancel()
+
+		n.queue.Push(queue.Message{
+			Revision: request.Revision,
+		})
+
+		// add another timeout since this receiver is called each time the channel has a message
+		n.addTimeout(ctx, selector)
+	})
+
+	n.addTimeout(ctx, selector)
+}
+
+func (n *NewRevision) addTimeout(ctx workflow.Context, selector workflow.Selector) {
+	n.timeout = false
+	ctx, cancel := workflow.WithCancel(ctx)
+	selector.AddFuture(workflow.NewTimer(ctx, n.timeoutDuration), func(f workflow.Future) {
+
+		// if canceled we shouldn't do anything
+		if ctx.Err() == workflow.ErrCanceled {
+			return
+		}
+
+		n.timeout = true
+	})
```

Comment:
Can we rename the newly defined ctx something like `ctxWithCancel` or `timeoutCtx` to distinguish it from the `addTimeout()` func parameter parent ctx?

From what I understand about this function, we create a child timeout ctx with a cancel function. Upon reaching the timeout/cancelling the pending timer we run the following behavior:

- if we canceled the pending timer, do nothing
- if we didn't, set timeout as true

Instead of setting timeout as true, should we be cancelling the parent context/shutting down the input receive channel? That way we stop trying to wait for new revisions to push to the queue and eventually cancel the workflow execution (at least until another revision restarts the whole process)?

Reply:
you're never supposed to close the channel at the receiver, it's the senders responsibility to do that. That logic is what I've used here as well.  Caller should handle timeouts and any result from that (in this case the selector).  In our specific case, we don't need to do anything special since this just is way for our selector to unblock and check the queue worker state.

Reply:
Also I don't want to create a different ctx value.  That leaves more opportunity to mess something up down the line and use the wrong ctx for something you didn't intend.